code/modules/vr/vr_avatar.dm

```diff
@@ -16,72 +17,120 @@
 	..()
 
 /mob/living/carbon/human/virtual_reality/death()
-	return_to_lobby()
+	myroom.players.Remove(src)
+	vr_all_players.Remove(src)
+	if((myroom.players.len == 0) && (myroom.expires == 1))
+		myroom.delete_timer = addtimer(myroom, "cleanup", 3 MINUTES)
+	else if((myroom.players.len == 1) && (myroom.expires == 2))
+		myroom.round_timer = addtimer(myroom, "vr_round", 10 SECONDS)
+		var/mob/living/carbon/human/virtual_reality/winner = myroom.players[1]
+		myroom.round_end = world.time + 1 MINUTES
+		for(var/mob/living/carbon/human/virtual_reality/P in vr_all_players)
+			if(P.ckey)
+				to_chat(P, "SYSTEM MESSAGE: [winner.name] has won the PVP match in The [myroom.name]. New match in one minute, please be in the lobby.")
+			else if(P.real_me.can_hear() && P.real_me.ckey && P in myroom.waitlist)
+				to_chat(P.real_me, "\[VR HEADSET\]: [winner.name] has won the PVP match in The [myroom.name]. New match in one minute, please be in the lobby.")
+
+	if(src.ckey)
+		return_to_lobby()
+	var/mob/living/carbon/human/virtual_reality/vr = src
+	var/list/corpse_equipment = vr.get_all_slots()
+	corpse_equipment += vr.get_equipped_items()
+	for(var/obj/O in corpse_equipment)
+		if(myroom.template.death_type == VR_DROP_ALL)
+			vr.unEquip(O)
+		else if(myroom.template.death_type == VR_DROP_BLACKLIST && !(O.type in myroom.template.drop_blacklist))
+			vr.unEquip(O)
+		else if(myroom.template.death_type == VR_DROP_WHITELIST && (O.type in myroom.template.drop_whitelist))
+			vr.unEquip(O)
+		else
+			qdel(O)
+	qdel(src)
 	return ..()
 
 /mob/living/carbon/human/virtual_reality/Destroy()
-	return_to_lobby()
 	return ..()
 
 /mob/living/carbon/human/virtual_reality/ghost()
-	set category = "OOC"
-	set name = "Ghost"
-	set desc = "Relinquish your life and enter the land of the dead."
-	var/mob/living/carbon/human/H = real_me
-	revert_to_reality(FALSE)
-	if(H)
-		H.ghost()
-	qdel(src)
+	alert(src, "You can not ghost while in VR. Please exit first.","Are you sure you want to ghost?")
+	return
 
-/mob/living/carbon/human/virtual_reality/proc/revert_to_reality(remove = 0)
-	if(real_me && mind)
-		mind.transfer_to(real_me)
+/mob/living/carbon/human/virtual_reality/say(var/message)
+	real_me.say(message)
+	..(message)
+
+/mob/living/carbon/human/virtual_reality/proc/revert_to_reality(var/remove)
+	if(real_me && ckey)
+		real_me.ckey = ckey
 		real_me.EyeBlind(2)
 		real_me.Confused(2)
+		var/mob/living/carbon/human/virtual_reality/vr = src
+		var/obj/item/clothing/ears/vr_headset/goggles
+		for(var/obj/item/clothing/ears/vr_headset/g in vr.real_me.contents)
+			goggles = g
+		if(goggles.contained() && vr_server_status == VR_SERVER_EMAG)
+			real_me.adjustBrainLoss(60)
+			real_me.vomit(20)
+		for(var/obj/item/clothing/ears/vr_headset/g in vr.real_me.contents)
```

Comment:
these types of variables are typically capitalized. although not sure that is a standard